### ISSUE TITLE
How to create grouped company-backends list to work with an extra one for specific major mode?

### ISSUE THREAD

**Issue by c02y** [company-mode/company-mode#585]
Lately, I'm working on how to configure the best grouped company-backends for different major modes, although having been using company-mode for over one year (already got a configuration long time ago), but recently realized it was not working as I expected, and started looking for the official documentation on how to configure it, but found nothing useful.

For example
in emacs-lisp-mode, words in buffers, yasnippet, elisp(company-capf), etc., sorted in this order in the droplist

in c/c++-mode, words in buffers, yasnippet, the name(variable, function, struct etc.), etc,.

in org-mode, words in buffers,  yasnippet, etc,

in latex-mode, words in buffers, yasnippet, latex-environment, etc.

If I insert and "/", it should use company-files as the primary backend.

In above examples, there will be a default company-backends for all the major modes when company-mode is enabled, there will be an extra configuration for a specific minor mode to work with the default.

For example:

    (setq company-backends
      '(
        (
         company-files
         company-dabbrev
         company-keywords
         company-capf
         company-yasnippet
         )
        (company-abbrev company-dabbrev)
        ))
    (add-hook 'emacs-lisp-mode-hook
		  (lambda ()
			(set (make-local-variable 'company-backends)
				 (append company-backends
						 '((
							company-elisp
							))))))

But I cannot make the company-elisp work, there always are such problems for other major modes, the two backends won't work well for a specific major mode. there always are one or more backends is not working

No official documentation or examples.

How to create grouped company-backends list to work with an extra one for specific major mode?

**Comment by dgutov** [comment]
You probably don't need to use `company-elisp` for Elisp: `company-capf` should suffice if you Emacs is modern enough (24.4+, IIRC).

Anyway, here. This is plain Emacs Lisp; there are documentation and tutorials for it. I don't feel any specific examples in Company's documentation are required.

```el
(add-hook 'emacs-lisp-mode-hook
          (lambda ()
            (set (make-local-variable 'company-backends)
                 (list
                  (cons 'company-elisp
                        (car company-backends))))))
```

If you really want `company-elisp` at the end, it will look something like this:

```el
(append (car company-backends)
        (list 'company-elisp))
```

**Comment by c02y** [comment]
Yes, indeed, I thought it didn't work because I typed "global-" or "global-s" or more, there will no candidates for "global-set-key", no droplist actually,  but I just type "global" or "globals" there will be candidates of "global-set-key", I guess it is the strategy of company-capf.

And I found that #350 use "(company-dabbrev-code company-dabbrev company-abbrev)" as a whole in the default company-backends, and what's the difference of those three? Do I need to use all three of them or only single one?

**Comment by dgutov** [comment]
> I thought it didn't work because I typed "global-" or "global-s" or more, there will no candidates for "global-set-key"

Did you have an opening paren before `global`?

> Do I need to use all three of them or only single one?

They're all different. Use the ones you need.

**Comment by c02y** [comment]
> Did you have an opening paren before global?
Yes, inside parens, if type directly inside **Scratch** buffer without parens, there will be no global-set-key candidate, even from company-capf.

**Comment by dgutov** [comment]
> Yes, inside parens

> without parens

I don't understand. Yes or no? When do you have the problem?

**Comment by c02y** [comment]
If without parens, there will be no candidate for global-set-key 

If with parens, there will be candidates for global-set-key if I type "global", but no candidates when I type "global-" or "global-s", eg., the company-elisp if not working.

Anyway, it is working now after I paste this part into my init.el:

    (add-hook 'emacs-lisp-mode-hook
              (lambda ()
                (set (make-local-variable 'company-backends)
                     (list
                      (cons 'company-elisp
                            (car company-backends))))))

Thank you.
